File: lib/_constraints.py
from typing import Callable, List, Union

from grilops import SymbolGrid
from grilops.geometry import Point, Direction
from grilops.regions import RegionConstrainer
from z3 import And, BoolRef, Implies, Not, Or, Sum

from lib._utils import junctions, sight_line, var


def no2x2(sg: SymbolGrid, good: Callable[[Point], Union[bool, BoolRef]]):
    for junction in junctions(sg):
        sg.solver.add(Not(And([good(p) for p in junction])))


def no_adjacent_symbols(sg: SymbolGrid, symbol: int, no_diagonal: bool = False):
    for p in sg.grid:
        for n in sg.vertex_sharing_neighbors(p) if no_diagonal else sg.edge_sharing_neighbors(p):
            sg.solver.add(Not(And(sg.cell_is(p, symbol), n.symbol == symbol)))


def no_adjacent_regions(sg: SymbolGrid, rc: RegionConstrainer):
    for p in sg.grid:
        for n in sg.edge_sharing_neighbors(p):
            sg.solver.add(
                Implies(
                    And(rc.region_id_grid[p] != -1, rc.region_id_grid[n.location] != -1),
                    rc.region_id_grid[p] == rc.region_id_grid[n.location],
                )
            )


def require_contiguous_block_sums(sg: SymbolGrid, line: List[Point], target_sums: List[Union[int, str]]):
    # For each horizontal/vertical line, use dynamic programming where num_blocks[i] is the current number of blocks
    # seen so far. Then anytime we end a block, we check if the block is the right sum and increment num_blocks[i] if
    # so. Otherwise, we keep num_blocks[i] as the same value.
    # If the current block sum is '*', then we also allow incrementing the block regardless of the current sum.
    if not target_sums:
        return
    num_blocks = [var() for _ in line]
    for i in range(1, len(line)):
        choices = [And(num_blocks[i] == num_blocks[i - 1], Or(sg.grid[line[i - 1]] == 0, Not(sg.grid[line[i]] == 0)))]
        for block_num, target_sum in enumerate(target_sums):
            if target_sum == "*":
                choices.append(And(num_blocks[i] == block_num + 1, num_blocks[i - 1] == block_num + 1))
            else:
                next_block_num = block_num + (
                    2 if block_num + 1 < len(target_sums) and target_sums[block_num + 1] == "*" else 1
                )
                for block_size in range(1, i):
                    squares = [sg.grid[line[i - j - 1]] for j in range(block_size)]
                    choices.append(
                        And(
                            num_blocks[i] == next_block_num,
                            num_blocks[i - block_size] == block_num,
                            sg.grid[line[i]] == 0,
                            *[square != 0 for square in squares],
                            True if target_sum == "?" else Sum(squares) == target_sum,
                            sg.grid[line[i - block_size - 1]] == 0
                        )
                    )
        sg.solver.add(Or(choices))
    sg.solver.add(num_blocks[0] == (1 if target_sums[0] == "*" else 0))
    sg.solver.add(num_blocks[-1] == len(target_sums))


def require_continuous(
    sg: SymbolGrid, good: Callable[[Point], Union[bool, BoolRef]], neighbors: Callable[[Point], List[Point]] = None
):
    tree = {p: var() for p in sg.grid}
    root = var()
    for p in sg.grid:
        sg.solver.add(
            Or(
                Not(good(p)),
                sg.lattice.point_to_index(p) == root,
                *[
                    And(good(q), tree[q] < tree[p])
                    for q in (neighbors(p) if neighbors else [n.location for n in sg.edge_sharing_neighbors(p)])
                ]
            )
        )


def require_loop_direction(sg: SymbolGrid):
    class DirectionRef:
        def __init__(self, dirs):
            self._directions = dirs
            self.ref = var()

        def __eq__(self, v: Direction):
            return self.ref == self._directions.index(v)

        def __ne__(self, v: Direction):
            return self.ref != self._directions.index(v)

    directions = sg.lattice.edge_sharing_directions()
    loop_direction = {p: DirectionRef(directions) for p in sg.grid}
    for p in sg.grid:
        sg.solver.add(And(loop_direction[p].ref >= -1, loop_direction[p].ref < len(directions)))
        sg.solver.add(Implies(loop_direction[p].ref == -1, Not(sg.symbol_set.is_loop(sg.grid[p]))))
        for v in directions:
            sg.solver.add(Implies(loop_direction[p] == v, sg.cell_is_one_of(p, sg.symbol_set.symbols_for_direction(v))))
            sg.solver.add(
                Implies(
                    loop_direction[p] == v,
                    loop_direction.get(p.translate(v)) != sg.lattice.opposite_direction(v),
                )
            )
    return loop_direction


def require_region_area(sg: SymbolGrid, start: Point, good: Callable[[Point], bool], target: int):
    # Perform a floodfill from the starting point to all contiguous shaded squares.
    # Ensure the total number of reached squares is equal to the number (plus one to include the original square).
    floodfill = {p: var() for p in sg.grid}
    for p in sg.grid:
        sg.solver.add(floodfill[p] == (start == p))
    for _ in range(min(target, len(sg.grid))):
        new_floodfill = {p: var() for p in sg.grid}
        for p in sg.grid:
            sg.solver.add(
                new_floodfill[p]
                == Or(
                    floodfill[p] == 1, *[And(good(p), floodfill[n.location] == 1) for n in sg.edge_sharing_neighbors(p)]
                )
            )
        floodfill = new_floodfill
    sg.solver.add(Sum([floodfill[p] for p in sg.grid]) == target)


def require_sight_line_count(sg: SymbolGrid, p: Point, good: Callable[[Point], bool], target: int):
    all_is_visible = [good(p)]
    for direction in sg.lattice.edge_sharing_directions():
        line = sight_line(sg, p, direction)
        for i in range(1, len(line)):
            all_is_visible.append(And([good(q) for q in line[: i + 1]]))
    sg.solver.add(Sum(all_is_visible) == target)

File: src/lib/constraints.ts
import { isEqual, range } from "lodash";
import { Arith, Bool, Solver } from "z3-solver";
import { ValueMap, ValueSet } from "./collections";
import { Bearing } from "./geometry/bearing";
import { Lattice } from "./geometry/lattice";
import { Point } from "./geometry/point";
import { PointSet } from "./geometry/pointset";
import { Vector } from "./geometry/vector";
import { LoopNetwork, Network, PathNetwork } from "./network";
import { Context } from "./solver";

export type ChoiceArith<T> = Arith & {
    is(val: T): Bool;
};

export type NetworkArith = Arith & {
    hasDirection(v: Vector): Bool;
    is(vs: Vector[]): Bool;
    isLoopSegment(): Bool;
    isStraight(): Bool;
    isTerminal(): Bool;
};

export interface Model {
    get(arith: Arith): number;
}

export interface Constraints {
    /** Creates a new variable that can take any integer value within the given bounds. */
    int(minInclusive?: number, maxInclusive?: number): Arith;

    /** Creates a new variable that can take any of the given values, or none of them (-1). */
    choice<V>(values: Iterable<V>): ChoiceArith<V>;

    /**
     * Returns a grid of variables that forms a valid network, e.g. if a cell goes right, then
     * the cell to its right also goes left. Each cell is a valid component of the given network.
     */
    NetworkGrid(points: PointSet, network: Network): ValueMap<Point, NetworkArith>;

    /**
     * Returns a valid network grid consisting of and only of non-intersecting paths.
     * @returns [network, grid, order] where order is -1 where there is no path, otherwise starts
     * from 0 and increments by 1 until the end.
     */
    PathsGrid(points: PointSet): [Network, ValueMap<Point, NetworkArith>, ValueMap<Point, Arith>];

    /**
     * Returns a valid network grid consisting of a single connected loop.
     * @returns [network, grid, root] where root is an arbitrary start point on the loop.
     */
    SingleLoopGrid(points: PointSet): [Network, ValueMap<Point, NetworkArith>, ChoiceArith<Point>];

    /** Adds the given constraints. */
    add(...exprs: Bool[]): void;

    /** Adds a constraint that all "good" points are orthogonally connected. */
    addAllConnected(points: PointSet, good: (p: Point) => Bool): void;

    /** Adds a constraint that all points are orthogonally connected along edges to one of the roots. */
    addConnected(
        points: PointSet,
        isRoot: (p: Point) => boolean | Bool,
        isEdge: (p: Point, q: Point) => boolean | Bool
    ): ValueMap<Point, ChoiceArith<Point>>;

    /**
     * Returns a variable representing the number of orthogonally connected points within the given
     * distance (inclusive) of the start point.
     */
    addContiguousArea(lattice: Lattice, points: PointSet, start: Point, good: (p: Point) => Bool, area: number): void;

    /** Adds a constraint that the contiguous sums of the given values (separated by 0) must have the given sums */
    addContiguousBlockSums(line: Arith[], sums: string[]): void;

    /** Adds a constraint that the orthogonally connected "good" points in straight lines have the given count. */
    addSightLineCount(
        lattice: Lattice,
        points: PointSet,
        start: Point,
        good: (p: Point, bearing?: Bearing) => Bool,
        count: number
    ): void;

    /**
     * Given the variables that we care about, returns the values for the variables that satisfy all added constraints.
     */
    solve(ariths: Iterable<[any, Arith]>): Promise<Model>;
}

export class ConstraintsImpl implements Constraints {
    constructor(
        private readonly differentFrom: ValueMap<any, number>,
        private readonly context: Context,
        private readonly solver: Solver,
        private readonly setValues: (values: ValueMap<any, number>) => void
    ) {}

    int(minInclusive?: number, maxInclusive?: number) {
        const arith = this.context.Int.const(crypto.randomUUID());
        if (minInclusive !== undefined) {
            this.solver.add(arith.ge(minInclusive));
        }
        if (maxInclusive !== undefined) {
            this.solver.add(arith.le(maxInclusive));
        }
        return arith;
    }

    choice<V>(valuesIt: Iterable<V>) {
        const values = [...valuesIt];
        return Object.assign(this.int(-1, values.length - 1), {
            is(val: V): Bool {
                return this.eq(values.findIndex(otherVal => isEqual(val, otherVal)));
            },
        });
    }

    NetworkGrid(points: PointSet, network: Network) {
        const { Implies, Or } = this.context;
        const grid = new ValueMap(points, p => {
            const directionSets = network.directionSets(p);
            const getIndices = (predicate: (vs: Vector[]) => Boolean) =>
                range(directionSets.length).filter(i => predicate(directionSets[i]));
            return Object.assign(this.int(0, network.directionSets(p).length - 1), {
                hasDirection(v: Vector) {
                    return Or(...getIndices(vs => vs.some(w => w.eq(v))).map(i => this.eq(i)));
                },
                is(vs: Vector[]) {
                    return Or(...getIndices(ws => isEqual(ws, vs.sort())).map(i => this.eq(i)));
                },
                isLoopSegment() {
                    return Or(...getIndices(vs => vs.length === 2).map(i => this.eq(i)));
                },
                isStraight() {
                    return Or(
                        ...getIndices(vs =>
                            points.lattice
                                .edgeSharingDirections(p)
                                .every(v => vs.some(w => w.eq(v)) === vs.some(w => w.eq(v.negate())))
                        ).map(i => this.eq(i))
                    );
                },
                isTerminal() {
                    return Or(...getIndices(vs => vs.length === 1).map(i => this.eq(i)));
                },
            });
        });
        for (const [p, arith] of grid) {
            for (const [q, v] of points.lattice.edgeSharingNeighbors(p)) {
                this.add(Implies(arith.hasDirection(v), grid.get(q)?.hasDirection(v.negate()) || false));
            }
        }
        return grid;
    }

    PathsGrid(points: PointSet): [Network, ValueMap<Point, NetworkArith>, ValueMap<Point, Arith>] {
        const { And, If, Implies, Or } = this.context;
        const network = new PathNetwork(points.lattice);
        const grid = this.NetworkGrid(points, network);
        const order = new ValueMap(points, _ => this.int());
        for (const [p, arith] of grid) {
            const hasNeighbor = (d: number) =>
                Or(
                    ...points
                        .edgeSharingNeighbors(p)
                        .map(([q, v]) => And(arith.hasDirection(v), order.get(q).eq(order.get(p).add(d))))
                );
            this.add(Implies(arith.isLoopSegment(), And(hasNeighbor(-1), hasNeighbor(1))));
            this.add(Implies(arith.isTerminal(), If(order.get(p).eq(0), hasNeighbor(1), hasNeighbor(-1))));
            this.add(Implies(arith.eq(0), order.get(p).eq(-1)));
        }
        return [network, grid, order];
    }

    SingleLoopGrid(points: PointSet): [Network, ValueMap<Point, NetworkArith>, ChoiceArith<Point>] {
        const { Or } = this.context;
        const network = new LoopNetwork(points.lattice);
        const grid = this.NetworkGrid(points, network);
        const root = this.choice(points);
        this.addConnected(
            points,
            p => Or(grid.get(p).eq(0), root.is(p)),
            (p, q) => grid.get(p).hasDirection(p.directionTo(q))
        );
        return [network, grid, root];
    }

    add(...exprs: Bool[]) {
        this.solver.add(...exprs);
    }

    addAllConnected(points: PointSet, good: (p: Point) => Bool) {
        const { And, Not, Or } = this.context;
        const tree = new ValueMap(points, _ => this.int());
        const root = this.choice(points);
        for (const p of points) {
            this.add(
                Or(
                    Not(good(p)),
                    root.is(p),
                    ...points.edgeSharingPoints(p).map(q => And(good(q), tree.get(q).lt(tree.get(p))))
                )
            );
        }
    }

    addConnected(
        points: PointSet,
        isRoot: (p: Point) => boolean | Bool,
        isEdge: (p: Point, q: Point) => boolean | Bool
    ) {
        const { And, If, Implies, Or } = this.context;
        const instance = new ValueMap(points, _ => this.choice(points));
        const tree = new ValueMap(points, _ => this.int());
        for (const p of points) {
            this.add(
                If(
                    isRoot(p),
                    instance.get(p).is(p),
                    Or(...points.edgeSharingPoints(p).map(q => And(isEdge(p, q), tree.get(q).lt(tree.get(p)))))
                )
            );
            for (const q of points.edgeSharingPoints(p)) {
                this.add(Implies(isEdge(p, q), instance.get(p).eq(instance.get(q))));
            }
        }
        return instance;
    }

    addContiguousArea(lattice: Lattice, points: PointSet, start: Point, good: (p: Point) => Bool, area: number) {
        const { And, Bool, Or, Sum } = this.context;
        let floodfill = new ValueMap(points, _ => Bool.fresh());
        for (const [p, arith] of floodfill) {
            this.add(arith.eq(p.eq(start) && good(p)));
        }
        for (let i = 0; i < area; i++) {
            const newFloodfill = new ValueMap(points, _ => Bool.fresh());
            for (const p of points) {
                this.add(
                    newFloodfill
                        .get(p)
                        .eq(
                            Or(
                                floodfill.get(p),
                                ...lattice.edgeSharingPoints(p).map(q => And(good(p), floodfill.get(q) || false))
                            )
                        )
                );
            }
            floodfill = newFloodfill;
        }
        this.add(Sum(...floodfill.values()).eq(area));
    }

    addContiguousBlockSums(line: Arith[], sums: string[]) {
        // For each horizontal/vertical line, use dynamic programming where num_blocks[i] is the
        // current number of blocks seen so far. Then anytime we end a block, we check if the block
        // is the right sum and increment num_blocks[i] if so. Otherwise, we keep num_blocks[i] as
        // the same value.
        // If the current block sum is '*', then we also allow incrementing the block regardless of
        // the current sum.
        const { And, Or, Sum } = this.context;
        if (sums.length === 0) {
            return;
        }
        line = [this.int(0, 0), ...line, this.int(0, 0)];
        const numBlocks = line.map(_ => this.int());
        for (let i = 1; i < line.length; i++) {
            const choices = [And(numBlocks[i].eq(numBlocks[i - 1]), Or(line[i - 1].eq(0), line[i].neq(0)))];
            for (const [blockNum, sum] of sums.entries()) {
                if (sum === "*") {
                    choices.push(And(numBlocks[i].eq(blockNum + 1), numBlocks[i - 1].eq(blockNum + 1)));
                } else {
                    const nextBlockNum = blockNum + (sums[blockNum + 1] === "*" ? 2 : 1);
                    for (let blockSize = 1; blockSize < i; blockSize++) {
                        const block = line.slice(i - blockSize, i);
                        choices.push(
                            And(
                                numBlocks[i].eq(nextBlockNum),
                                numBlocks[i - blockSize].eq(blockNum),
                                line[i].eq(0),
                                ...block.map(arith => arith.neq(0)),
                                sum === "?" ? true : Sum(...block).eq(parseInt(sum)),
                                line[i - blockSize - 1].eq(0)
                            )
                        );
                    }
                }
            }
            this.add(Or(...choices));
        }
        this.add(numBlocks[0].eq(sums[0] === "*" ? 1 : 0));
        this.add(numBlocks[numBlocks.length - 1].eq(sums.length));
    }

    addSightLineCount(
        lattice: Lattice,
        points: PointSet,
        start: Point,
        good: (p: Point, bearing?: Bearing) => Bool,
        count: number
    ) {
        const { And, Sum } = this.context;
        const cells = new ValueSet([start]);
        const visible = [good(start)];
        for (const bearing of lattice.bearings()) {
            const line = points.lineFrom(start, bearing);
            for (let i = 1; i < line.length; i++) {
                if (!cells.has(line[i])) {
                    cells.add(line[i]);
                    visible.push(And(...line.slice(0, i + 1).map(p => good(p, bearing))));
                }
            }
        }
        this.add(Sum(...visible).eq(count));
    }

    async solve(arithsIt: Iterable<[any, Arith]>) {
        const ariths = [...arithsIt];
        if (this.differentFrom) {
            // uniqueness constraint
            this.solver.add(this.context.Or(...ariths.map(([key, arith]) => arith.neq(this.differentFrom.get(key)))));
        }
        switch (await this.solver.check()) {
            case "sat":
                const z3Model = this.solver.model();
                const model = { get: (arith: Arith) => Number((z3Model.eval(arith) as any).value()) };
                this.setValues(new ValueMap(ariths.map(([key, arith]) => [key, model.get(arith)])));
                return model;
            case "unsat":
                throw new Error(this.differentFrom ? "No other solution found." : "No solution found.");
            case "unknown":
                throw new Error("An unknown error occurred.");
        }
    }
}
